src/ray/common/memory_monitor.cc

```diff
@@ -0,0 +1,185 @@
+#include <tuple>
+
+#include "ray/common/nullable_uint.h"
+#include "ray/common/ray_config.h"
+#include "ray/util/logging.h"
+
+namespace ray {
+
+MemoryMonitor::MemoryMonitor(float usage_threshold_,
+                             uint64_t monitor_interval_ms,
+                             MemoryUsageRefreshCallback monitor_callback)
+    : usage_threshold_(usage_threshold_),
+      monitor_callback_(monitor_callback),
+      io_context_(),
+      monitor_thread_([this] {
+        boost::asio::io_service::work io_service_work_(io_context_);
+        io_context_.run();
+      }),
+      runner_(io_context_) {
+  RAY_CHECK(monitor_callback_ != nullptr);
+  RAY_CHECK_GE(usage_threshold_, 0);
+  RAY_CHECK_LE(usage_threshold_, 1);
+  if (monitor_interval_ms > 0) {
+#ifdef __linux__
+    runner_.RunFnPeriodically(
+        [this] {
+          bool is_usage_above_threshold = IsUsageAboveThreshold();
+          monitor_callback_(is_usage_above_threshold);
+        },
+        monitor_interval_ms,
+        "MemoryMonitor.CheckIsMemoryUsageAboveThreshold");
+    RAY_LOG(INFO) << "MemoryMonitor initialized";
+#else
+    RAY_LOG(WARNING) << "Not running MemoryMonitor. It is currently supported "
+                     << "only on Linux.";
+#endif
+  } else {
+    RAY_LOG(INFO) << "MemoryMonitor disabled. Specify "
+                  << "`memory_monitor_interval_ms` > 0 to enable the monitor.";
+  }
+}
+
+bool MemoryMonitor::IsUsageAboveThreshold() {
+  auto [used_memory_bytes, total_memory_bytes] = GetMemoryBytes();
+  if (total_memory_bytes == kNull || used_memory_bytes == kNull) {
+    RAY_LOG(WARNING) << "Unable to capture node memory. Monitor will not be able "
+                     << "to detect memory usage above threshold.";
+    return false;
+  }
+  auto usage_fraction = static_cast<float>(used_memory_bytes) / total_memory_bytes;
+  bool is_usage_above_threshold = usage_fraction > usage_threshold_;
+  if (is_usage_above_threshold) {
+    RAY_LOG_EVERY_MS(INFO, 1000)
+        << "Node memory usage above threshold, used: " << used_memory_bytes
+        << ", total: " << total_memory_bytes << ", usage fraction: " << usage_fraction
+        << ", threshold: " << usage_threshold_;
+  }
+  return is_usage_above_threshold;
+}
+
+std::tuple<nuint64_t, nuint64_t> MemoryMonitor::GetMemoryBytes() {
+  RAY_CHECK(__linux__) << "Memory monitor currently supports only linux";
+  auto [system_used_bytes, system_total_bytes] = GetLinuxMemoryBytes();
+  auto [cgroup_used_bytes, cgroup_total_bytes] = GetCGroupMemoryBytes();
+  system_total_bytes = NullableInt::min(system_total_bytes, cgroup_total_bytes);
+  if (system_total_bytes == cgroup_total_bytes) {
+    system_used_bytes = cgroup_used_bytes;
+  }
+  return std::tuple(system_used_bytes, system_total_bytes);
+}
+
+std::tuple<nuint64_t, nuint64_t> MemoryMonitor::GetCGroupMemoryBytes() {
+  nuint64_t total_bytes = kNull;
+  if (boost::filesystem::exists(kCgroupsV2MemoryMaxPath)) {
```

Comment:
Just std::filesystem::exists since we are using c++17.

Reply:
sg